sensor/common/compliance/nodescan_handler_test.go

```diff
@@ -0,0 +1,161 @@
+package compliance
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	timestamp "github.com/gogo/protobuf/types"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/sync"
+	scannerV1 "github.com/stackrox/scanner/generated/scanner/api/v1"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestNodeScanHandler(t *testing.T) {
+	suite.Run(t, new(NodeScanHandlerTestSuite))
+}
+
+func fakeNodeScanV2(nodeName string) *storage.NodeScanV2 {
+	msg := &storage.NodeScanV2{
+		NodeId:   "",
+		NodeName: nodeName,
+		ScanTime: timestamp.TimestampNow(),
+		Components: &scannerV1.Components{
+			Namespace: "Testme OS",
+			RhelComponents: []*scannerV1.RHELComponent{
+				{
+					Name:      "vim-minimal",
+					Namespace: "rhel:8",
+					Version:   "2:7.4.629-6.el8.x86_64",
+					Arch:      "x86_64",
+					Module:    "FakeMod",
+					Cpes:      []string{"cpe:/a:redhat:enterprise_linux:8::baseos"},
+					AddedBy:   "FakeLayer",
+				},
+			},
+			LanguageComponents: nil,
+		},
+		Notes: nil,
+	}
+	return msg
+}
+
+type nodeScansProducerThatStops struct {
+	numMsgs   int
+	nodeScans chan *storage.NodeScanV2
+}
+
+type stoppable interface {
+	Stop(error)
+}
+
+func newNodeScansProducerThatStops(num int) *nodeScansProducerThatStops {
+	return &nodeScansProducerThatStops{
+		numMsgs:   num,
+		nodeScans: make(chan *storage.NodeScanV2),
+	}
+}
+
+func (nsp *nodeScansProducerThatStops) generator() <-chan *storage.NodeScanV2 {
+	return nsp.nodeScans
+}
+
+func (nsp *nodeScansProducerThatStops) start(s stoppable) {
+	go func(numMsgs int) {
+		defer close(nsp.nodeScans)
+		for i := 0; i < numMsgs; i++ {
+			nsp.nodeScans <- fakeNodeScanV2(fmt.Sprintf("Bobby_%d", i))
+			if i == 0 {
+				s.Stop(nil)
+			}
+		}
+	}(nsp.numMsgs)
+}
+
+type NodeScanHandlerTestSuite struct {
+	suite.Suite
+	mu                   *sync.Mutex
+	numReceivedAtCentral int
+	cancel               context.CancelFunc
+	ctx                  context.Context
+	toCentral            chan *central.MsgFromSensor
+}
+
+func (s *NodeScanHandlerTestSuite) SetupTest() {
+	s.mu = &sync.Mutex{}
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
Curious: why locking mutex in `SetupTest()`?

Reply:
Only to prevent race in accessing `numReceivedAtCentral`. I will try to optimize it out, as the mutex can hide some other potential races.

Reply:
The mutex and the counter var `numReceivedAtCentral` are gone now.